### PR TITLE
fix(runner): long synchronous tasks does not time out (fix #2920)

### LINKED ISSUE #2920: Long task with synchronous pieces does not time out
### Describe the bug

Basically I've got a synchronous function that I want to test is reasonably fast with a large amount of data. I understand the issue of not being able to interrupt the process for the synchronous task, and saw for example another issue mentioning the problem with an infinite loop—but this function does eventually resolve. If I've set a timeout for 150ms and it takes 5 seconds to resolve, I would expect the test to fail as timed out, but it doesn't, no matter how long it takes.

I suppose this may be "behaving as expected" as timeout is _only_ intended to interrupt asynchronous tasks if it can and does nothing if it can't interrupt, but it's definitely unintuitive to me that a test with a 150ms timeout that resolves in 5 seconds is considered to have passed.

For my particular use case I did a `const start = Date.now(); … expect(Date.now() - start).toBeLessThan(150);` and I'm sure that will probably be the response here, but I think it's unintuitive that the distinction between synchronous and asynchronous code (even synchronous code within a promise) will change whether the same behavior is considered to have timed out.

### Reproduction

Times out as I'd expect, because (as far as I understand, and I admit I'm a little fuzzy) `setTimeout` does not monopolize ticks and allows the test process to intervene:

```ts
it('times out', async () => {
  await new Promise((resolve) => {
    setTimeout(resolve, 5_000);
  });
}, 150);
```

Does not time out, and does not _fail_ once completed:

```ts
it('does not time out', async () => {
  const start = Date.now();
  await new Promise((resolve) => {
    while (Date.now() - start < 5_000) {}
    resolve();
  });
}, 150);
```

Even this, which occurred to me while writing this issue, does not result in a timeout:

```ts
it('does not time out even with a promise after completion allowing the test runner to intervene', async () => {
  const start = Date.now();
  while (Date.now() - start < 5_000) {}
  await new Promise((resolve) => {
    setImmediate(resolve); // setTimeout(resolve, 0 or even 10) has no effect either
  });
}, 150);
```

### System Info

```shell
System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 85.13 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.10.0 - ~/.nodenv/versions/18.10.0/bin/node
    npm: 8.19.2 - ~/.nodenv/versions/18.10.0/bin/npm
  Browsers:
    Firefox: 109.0.1
    Safari: 16.2
  npmPackages:
    @vitest/coverage-c8: ^0.24.5 => 0.24.5 
    vite: ^3.2.4 => 3.2.4 
    vitest: ^0.24.5 => 0.24.5
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by ferdodo**
### Description

Fixes #2920 

This PR solves Vitest ignoring timeouts on long synchronous tasks, that are blocking until the task and the timeout are settled simultaneously.

In this case we want to give priority to the timeout, because a test running longer than it's own timeout shall fail.

- The result of the test is wrapped into a promise, and delayed to be processed in the same event loop cycle as the timeout.
- I changed the order in the Promise.race, to give the timeout priority over the other promise.
- I added a test for this case that takes 15ms, both changes are necessary to make the test pass.
- Using nextTick from getSafeTimers() over the additional setTimeout I used did not solve the issue.
- Outputing logs during tests displays slighly differently.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by ferdodo** [comment]
fixes #6944

File: packages/runner/src/context.ts [hunk 1]
```diff
@@ -52,6 +51,9 @@ export function withTimeout<T extends (...args: any[]) => any>(
         // `unref` might not exist in browser
         timer.unref?.()
       }),
+      Promise.resolve(fn(...args)).then((result) => {
+        return new Promise(resolve => setTimeout(resolve, 0, result))
+      }),
```
Comment (hi-ogawa) [inline packages/runner/src/context.ts:56]: I took a while to understand this but the idea is more or less like this? Executing `fn` in sync as usual, but holding off returning until `setTimeout(.., 0)` to flush other `setTimeout` callbacks. I think code is fine as is, but it's just my commentary.

```ts
      (async () => {
        const result = await fn();
        await new Promise(resolve => setTimeout(resolve, 0))
        return result;
      })()
```

**Review by hi-ogawa — APPROVED** [review]
Nice find! The change makes sense to me.

File: packages/runner/src/context.ts (hunk 1 above)
Comment (ferdodo) [reply packages/runner/src/context.ts:56]: Thanks a lot! Yes that's the same thing, I can try refactor with your code if it's best.

**Comment by ferdodo** [comment]
The Browser: webkit test was passing in one of my previous run: 

https://github.com/vitest-dev/vitest/actions/runs/11955759654

Is it a flakiness issue ? It might just need a job re-run to pass.

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**